### DISCUSSION TITLE
custom-style support for tables in ConTeXt writer

### DISCUSSION THREAD

**Discussion by massifrg** [jgm/pandoc#10208] [Ideas]
This is a feature request to have `\startxtable[myCustomStyle]` instead of simply `\startxtable` in ConTeXt output, when a `Table` block has a `custom-style` attribute (in the example its value is "myCustomStyle").

---

[Pandoc 3.4](https://pandoc.org/releases.html#pandoc-3.4-2024-09-09) added support for `custom-style` attribute in Docx writer.

Such attribute could be used in ConTeXt output too.

Pandoc's ConTeXt writer outputs a table as:

```tex
\startxtable
...
\stopxtable
```

When the table has a `custom-style` attribute, e.g. with value "myStyle", the output could be:

```tex
\startxtable[myStyle]
...
\stopxtable
```

so that you could customize the table with these ConTeXt commands:

```tex
\definextable[myStyle]
\setupxtable[myStyle][...]
```

These custom definitions could be written in a user-defined template, or in a `RawBlock` of type `context` preceding the `Table` in the document.

Without the definitions, `\startxtable[myStyle]` would be equivalent to `\startxtable` in ConTeXt output: you would just get some error messages in its log, but the table would be there, just not customized.

For further reference, you can see the [Extreme tables](https://www.pragma-ade.nl/general/manuals/xtables-mkiv.pdf) manual and [\setupxtable](https://wiki.contextgarden.net/Command/setupxtable) in the ConTeXt WIKI.

**Comment by jgm** [comment]
That seems fairly sensible, but it would be a bit weird if we enabled the `custom_styles` extension for ConTeXt and it only affected tables.  Is there a similarly sensible thing to do with custom-style attributes on other elements?

**Comment by badumont** [comment]
It could be used with Divs and Spans too. A Div with custom_styles="mystyle" could surround its content with \startmystyle … \stopmystyle. A Span with the same attribute would place its content in the unique argument of the macro \mystyle. Both \mystyle and  \startmystyle … \stopmystyle could be defined by the user with \definestartstop[mystyle][...=...,...] (examples here: https://wiki.contextgarden.net/Command/definestartstop).

**Reply by jgm** [reply]
But in this case an error would be raised and processing would stop if the special style is not defined?

**Reply by massifrg** [reply]
Yes, that's the problem.

Pandoc should put a `\definestartstop[mystyle]` at the start of the document, leaving to the user to "fill" it with `\setupmystyle[...=..., ....]` in the template or inside a RawBlock.

Pandoc already does something conceptually similar with DOCX, ICML and ODT, creating empty styles with the names provided to the `custom-style` attributes.

There are two problems with `\definestartstop` though:
- there could be clashes with existing `\start ... \stop` environments (e.g. "narrower")
- the ConTeXt WIKI says that `\definestartstop` is suitable at block level (so it would be only for paragraphs and not spans), but I'm not sure that's true

I'm asking to the ConTeXt mailing list for the best ways to express Span and paragraphs (Div) custom styles.

I have some ideas -- i.e. `\definehighlight` for Spans and paragraph setups for Divs -- but better ask someone with more knowledge.

**Comment by badumont** [comment]
> I'm asking to the ConTeXt mailing list for the best ways to express Span
> and paragraphs (Div) custom styles.

As I wrote, \definestartstop[mystyle] can define both \startmystyle … \stopmystyle (block-level) and \mystyle{…} (inline). There is an example of that in the page linked above.

> Pandoc should put a \definestartstop[mystyle] at the start of the
> document, leaving to the user to "fill" it with
> \setupmystyle[...=..., ....] in the template or inside a RawBlock.

I think that it is good enough.

> -   there could be clashes with existing \start ... \stop environments
>     (e.g. "narrower")

Sure, but it does not seem a problem to me to let the user check this as he decides to use custom styles instead of the default constructs provided by Pandoc. This is a bit like RawInlines giving the user the power to generate an invalid document.

**Reply by massifrg** [reply]
> Sure, but it does not seem a problem to me to let the user check this as he decides to use custom styles instead of the default constructs provided by Pandoc. This is a bit like RawInlines giving the user the power to generate an invalid document.

When you write a RawInline, you know you are messing with low level, potentially dangerous stuff.
It's not so apparent when you're just naming a custom style.

Anyway, there's a workaround: adding a prefix in ConTeXt.
That way `custom-style="myStyle"` could become `\definestartstop[CSmyStyle]`, `\startCSmyStyle ... \stopCSmyStyle` and `\CSmyStyle{ ... }` in ConTeXt.

BTW, here's the link to [the thread in ConTeXt mailing list](https://mailman.ntg.nl/archives/list/ntg-context@ntg.nl/thread/QSO6NZSYIBT6ZMKSFM7OPNKHFAQ7NZGX/).

**Comment by bpj** [comment]
This would be useful for LaTeX too.

 Not perhaps for tables (are there any packages for tables with "styles"?) but definitely for Divs/environments and Spans/commands.

Of course it should be subject to an extension which is off by default. This probably goes without saying but in case it does not...

I'm (unsurprisingly) currently doing this with filters, although I call my attributes `cmd` and `env` because there are some commands which act on paragraphs and some environments which work inline. While you probably can wrap the former in an environment you can't TTBOMK (always) wrap the latter in a command.

I even have some filters which use mini-templates (defined as Code(Block)s in metadata or in external files) so that you can interpolate values from attributes or (a subtree of) metadata and insert a RawBlock/RawInline with the result before and/or after the Div/Span. One version uses a fairly "sophisticated" library using Lpeg/re; the others uses a simple `string.gsub` with a table or function as replacement argument. (You can't use an Attr as replacement argument, and copying it into a table is more complicated than it should be so a function is actually more effective.)

Ideally Pandoc should make arguments available as key—value pairs to TeX, at least such with some prefix. There is the question of how to handle that in LaTeX. There is the *keycommand* package, but it has been unmaintained for a long time and AFAIK still is.

**Comment by badumont** [comment]
> This would be useful for LaTeX too.
> 
> Not perhaps for tables (are there any packages for tables with
> "styles"?) but definitely for Divs/environments and Spans/commands.
> 
> Of course it should be subject to an extension which is off by default.
> This probably goes without saying but in case it does not...

Sure!

> Ideally Pandoc should make arguments available as key—value pairs to
> TeX, at least such with some prefix. There is the question of how to
> handle that in LaTeX. There is the keycommand package, but it has
> been unmaintained for a long time and AFAIK still is.

https://ctan.org/pkg/expkv-bundle should work with all TeX formats.

**Comment by massifrg** [comment]
The implementation for ConTeXt could be this:

```tex
...\start[myStyle]text with character custom style\stop{}...
```

for `Span`s with `custom-style`. Please notice the braces `{}` after `\stop`, otherwise the possible, following space would disappear in the output.

For `Div`s, the implementation could be like this:

```tex
...
\start[myStyle]
paragraphs (blocks) with custom paragraph style
\stop
...
```

That works without errors even when there's no `\definestartstop[myStyle]` before;
the effect is the same of a group: `{...}` or `\bgroup ...\egroup` in TeX/ConTeXt terms.

`\startmyStyle` would give an error without a preceding `\definestartstop[myStyle]`, `\start[myStyle]` does not.
Using brackets lets you use names with spaces too, e.g. `\start[My Style]`.

For simplicity, I'd follow @badumont advice of leaving to the user to check clashes with existing ConTeXt environments;
In the documentation, I'd just give the advice to prefer names with a capital letter for custom styles, to prevent clashes, since predefined environments are all lowercase.

I thank Wolfgang Schuster of the ConTeXt Group for the advice he gave me on this topic:

> You can try the the \start command which accepts an optional argument which expands to the environment given by the argument when available.
> 
> When there is no environment with the given name the command pair results in a local group for the content between them.
> 
> In the following example the first line ignores the argument of \start because no \startimportant environment exists at this point. The second line on the other hand gets formatted because I used \definestartstop to define the \startimportant environment.
> 

```tex
\starttext

before \start[important]inbetween\stop{} after

\definestartstop [important] [color=red,style=bold]

before \start[important]inbetween\stop{} after

\stoptext
```

**Reply by massifrg** [reply]
There's a -- acceptable, IMHO -- side effect on `Div`s with a `custom-style`: they are used in Pandoc to customize the paragraphs they contain, resulting in custom paragraph styles in DOC, ICML, ODT exports.

In ConTeXt, they'd have a wider effect:

```tex
\definestartstop [important] [color=red,style=bold]
\start[important]
...
\stop
```

would make all the stuff between `\start[important]` and `\stop` red and bold: not just paragraphs , also tables, bullet lists, etc.
